### ISSUE TITLE
python does not have cPickle module on SuSe13

### ISSUE THREAD

**Issue by davydden** [LLNL/spack#2253]
**update:**
it looks like there is a problem with `python` built by spack, see [below](https://github.com/LLNL/spack/issues/2253#issuecomment-259417869).

---
those who use python, do you know what could be the problem with `cPickle` module missing in Python built by Spack:
```
cat /tmp/iwtm009/spack-stage/spack-stage-mrG47q/petsc-3.7.4/spack-build.out
<bound method Petsc.install of <spack.pkg.builtin.petsc.Petsc object at 0x71e99d0>>
==> './configure' '--prefix=/calculate/spack/opt/spack/linux-SuSE13-x86_64/gcc-4.8/petsc-3.7.4-twnt2hfl2dulrsetdkr4ulfvnkh43tkq' '--with-ssl=0' '--download-c2html=0' '--download-hwloc=0' '--with-mpi=1' '--with-mpi-dir=/calculate/spack/opt/spack/linux-SuSE13-x86_64/gcc-4.8/openmpi-2.0.1-osy2cutkplq4i4iwbam2etf573axq3bv' '--with-cpp=cpp' '--with-cxxcpp=cpp' '--with-precision=double' '--with-scalar-type=real' '--with-shared-libraries=1' '--with-debugging=0' '--with-blas-lapack-lib=/calculate/spack/opt/spack/linux-SuSE13-x86_64/gcc-4.8/openblas-0.2.19-td2nztpmot3lngv4cmistsibu3p6zcx6/lib/libopenblas.so' '--with-metis=1' '--with-metis-dir=/calculate/spack/opt/spack/linux-SuSE13-x86_64/gcc-4.8/metis-5.1.0-uri2t4x6h26d4q3rpf4cjln46dw3cmzv' '--with-boost=0' '--with-hdf5=1' '--with-hdf5-dir=/calculate/spack/opt/spack/linux-SuSE13-x86_64/gcc-4.8/hdf5-1.10.0-patch1-r5lcom3ltwcc3jdpmk2dlf4zy4ntaz5h' '--with-hypre=1' '--with-hypre-dir=/calculate/spack/opt/spack/linux-SuSE13-x86_64/gcc-4.8/hypre-2.11.1-d72uxnfitqiybgdvl3a5hxusrvq3u3vq' '--with-parmetis=1' '--with-parmetis-dir=/calculate/spack/opt/spack/linux-SuSE13-x86_64/gcc-4.8/parmetis-4.0.3-bbpmfqbjcyufac3mtbjmiwghxausiydj' '--with-mumps=1' '--with-mumps-dir=/calculate/spack/opt/spack/linux-SuSE13-x86_64/gcc-4.8/mumps-5.0.2-m3bcskvwe5hnrkylnwsklmre7qpsibpf' '--with-scalapack=1' '--with-scalapack-dir=/calculate/spack/opt/spack/linux-SuSE13-x86_64/gcc-4.8/netlib-scalapack-2.0.2-l5n4cd5yctjn4zi2s5obb2lpn4tvbhqm' '--with-superlu_dist-include=/calculate/spack/opt/spack/linux-SuSE13-x86_64/gcc-4.8/superlu-dist-5.1.1-xhleqgylzdhciku66dhro2s7d6k6vz24/include' '--with-superlu_dist-lib=/calculate/spack/opt/spack/linux-SuSE13-x86_64/gcc-4.8/superlu-dist-5.1.1-xhleqgylzdhciku66dhro2s7d6k6vz24/lib/libsuperlu_dist.a' '--with-superlu_dist=1'
===============================================================================
            Configuring PETSc to compile on your system
===============================================================================
Traceback (most recent call last):
 File "./configure", line 10, in <module>
   execfile(os.path.join(os.path.dirname(__file__), 'config', 'configure.py'))
 File "./config/configure.py", line 496, in <module>
   petsc_configure([])
 File "./config/configure.py", line 396, in petsc_configure
   import config.base
 File "/tmp/iwtm009/spack-stage/spack-stage-mrG47q/petsc-3.7.4/config/BuildSystem/config/base.py", line 60, in <module>
   import script
 File "/tmp/iwtm009/spack-stage/spack-stage-mrG47q/petsc-3.7.4/config/BuildSystem/script.py", line 6, in <module>
   import cPickle
ImportError: No module named cPickle
```
?

Do you think it's PETSc issue or Python built by Spack? Anyone tried bulding petsc with Spack on SuSE13?

**Comment by citibeth** [comment]
It looks like PETSc is trying to build with Python3 instead of its required Python2.  This could potentially be coming from your Spack (if you built Python3 and then loaded it into your environment and you're building with `--dirty`).  Or more likely from your SuSE.  Linux distros are just beginning to ship Python3 standard, the phase-out of Python2 continues...

If you build the standard Python tarball (which Spack does), then Python2 will install binaries called `python` and `python2`, and Python3 will install a binary called `python3`.  But not all distros respect this.  Macintosh installs Python2 under `python`, but does not install `python2` at all.  I believe Anaconda on Windows installs `python`, even when Python3 is being installed.  The end result is a big total mess.

Bottom line is, you need to find a Python2 to use to configure PETSc.  (PETSc makes a "configure" script, but it's actually a Python2 script, not a GNU autotools thing).  See the commit that temporarily addressed that problem for me on my system: https://github.com/citibeth/spack/commit/6bbd6a510d2999b8ba4799a2212e5ee8bf09ad68#diff-b5b58063038a12c77d694aea3a86f840

(I removed Python as a dependency in that commit to get around #2098; I assume this is not a problem for you.  But the other part of the commit might be helpful).

https://github.com/citibeth/spack/commit/6bbd6a510d2999b8ba4799a2212e5ee8bf09ad68#diff-b5b58063038a12c77d694aea3a86f840

**Comment by davydden** [comment]
Thanks @citibeth , i will look closer into what is happening and especially how `python3` could possibly be picked up. The only recent change to `python` package is https://github.com/LLNL/spack/pull/2173 , maybe it is somewhat related... I will comment later in this thread once I know more...

**Comment by citibeth** [comment]
I added Python2 to my `packages.yaml`.  No point in building Python just as
a build dependency for other stuff.  If Spack is running, then by
definition you ALREADY have a Python 2.6 or 2.7 that should be suitable for
building PETSc.

@tgamblin

One idea for these things that require Python as build dependency... would
be to able to just ask Spack for its Python, and use that to run the
`configure` script.  This procedure would be robust, and would always work,
and wouldn't run into concretization bugs with build dependencies.  Then
`python` wouldn't have to be listed as a dependency at all.  BUT... it
feels like a hack.  Todd, your thoughts?

**Comment by davydden** [comment]
Workaround like
```
which('python2')('config/configure.py', '--prefix=%s' % prefix, *options)
```
did not help. So it looks like `python@2.7.12` build by `Spack` somehow lacks this module for whatever reason.

**Comment by citibeth** [comment]
I would try some experimentation.  Run the Python you think it's getting
and try 'import cPickle'.  This is really strange, `cPickle' is a standard
part of Python2.

Or... just use your System Python.

**Comment by christianbaensch** [comment]
Here is the error message when we try using python from Spack
```
# /calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/python-2.7.12-yotbp4vq4sikgho7gxdoadaf7j3kr2zd/bin/python -c 'import cPickle'
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: No module named cPickle
```

**Comment by alalazo** [comment]
@christianbaensch @davydden I think you need to inspect closely the `build.out` of `python`. I encountered a similar case where the `urllib` module in a `spack` built python was severely flawed on a `rhel`  machine because some `*-devel` system package was missing.

**Comment by davydden** [comment]
> ...because some *-devel system package was missing.

does not this mean that we are missing some dependencies in `python`, i.e. rely on some additional system packages to be available?

**Comment by citibeth** [comment]
What OS are you on?

Again... If you just want petsc, use srstem python for this build dep.

**Comment by davydden** [comment]
@citibeth : `SuSE13` (see logs above), it's not about workarounds, I know that it is possible. I just want to find the source of the problem because all works fine on Ubuntu14,16 and macOS. I would prefer to have this resolved before `0.10` release.

@christianbaensch : could you check `PYTHONHOME` variable in `spack-build.env` in temp build folder of PETSc? On Ubuntu it points for me to the python prefix in Spack, i.e. `.../opt/spack/linux-ubuntu16-x86_64/python-2.7.12-abracadabra`.

**Comment by davydden** [comment]
@alalazo : I checked build logs of Python, here are a few things (full logs are attached)
```
configure: loading site script /usr/share/site/x86_64-unknown-linux-gnu
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
...
/calculate/spack/lib/spack/env/gcc/gcc -pthread -c -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes  -I. -IInclude -I./Include -I/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/bzip2-1.0.6-qzrsmlmxgad7ckz7ob2xnz3ir7wchclp/include -I/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/ncurses-6.0-vu7i5i6w5d22kkb3dn2i3xm3bnwmayvv/include -I/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/zlib-1.2.8-qsdhmcv7nngu2djw7jvldxzb667sxxai/include -I/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/openssl-1.0.2j-nf26w5xnlarn7b4tezfsqpndhsx6coa5/include -I/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/sqlite-3.8.5-oxvmqedc2fcqgv4c6xzullyibpnke3hk/include -I/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/readline-6.3-4rrhdjo2lectv5nyyai3cad6cu4zjepd/include -fPIC -DPy_BUILD_CORE \
      -DSVNVERSION="\"`LC_ALL=C svnversion .`\"" \
      -DHGVERSION="\"`LC_ALL=C hg id -i .`\"" \
      -DHGTAG="\"`LC_ALL=C hg id -t .`\"" \
      -DHGBRANCH="\"`LC_ALL=C hg id -b .`\"" \
      -o Modules/getbuildinfo.o ./Modules/getbuildinfo.c
ImportError: No module named site
ImportError: No module named site
ImportError: No module named site
```
but otherwise I don't see anything suspicious. `cPickle` seems to be built
```
building 'cPickle' extension
/calculate/spack/lib/spack/env/gcc/gcc -pthread -fPIC -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I. -IInclude -I./Include -I/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/bzip2-1.0.6-qzrsmlmxgad7ckz7ob2xnz3ir7wchclp/include -I/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/ncurses-6.0-vu7i5i6w5d22kkb3dn2i3xm3bnwmayvv/include -I/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/zlib-1.2.8-qsdhmcv7nngu2djw7jvldxzb667sxxai/include -I/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/openssl-1.0.2j-nf26w5xnlarn7b4tezfsqpndhsx6coa5/include -I/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/sqlite-3.8.5-oxvmqedc2fcqgv4c6xzullyibpnke3hk/include -I/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/readline-6.3-4rrhdjo2lectv5nyyai3cad6cu4zjepd/include -I/usr/local/include -I/tmp/iwtm009/spack-stage/spack-stage-sP5iP4/Python-2.7.12/Include -I/tmp/iwtm009/spack-stage/spack-stage-sP5iP4/Python-2.7.12 -c /tmp/iwtm009/spack-stage/spack-stage-sP5iP4/Python-2.7.12/Modules/cPickle.c -o build/temp.linux-x86_64-2.7/tmp/iwtm009/spack-stage/spack-stage-sP5iP4/Python-2.7.12/Modules/cPickle.o
/calculate/spack/lib/spack/env/gcc/gcc -pthread -shared -L/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/bzip2-1.0.6-qzrsmlmxgad7ckz7ob2xnz3ir7wchclp/lib -L/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/ncurses-6.0-vu7i5i6w5d22kkb3dn2i3xm3bnwmayvv/lib -L/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/zlib-1.2.8-qsdhmcv7nngu2djw7jvldxzb667sxxai/lib -L/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/openssl-1.0.2j-nf26w5xnlarn7b4tezfsqpndhsx6coa5/lib -L/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/sqlite-3.8.5-oxvmqedc2fcqgv4c6xzullyibpnke3hk/lib -L/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/readline-6.3-4rrhdjo2lectv5nyyai3cad6cu4zjepd/lib build/temp.linux-x86_64-2.7/tmp/iwtm009/spack-stage/spack-stage-sP5iP4/Python-2.7.12/Modules/cPickle.o -L/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/bzip2-1.0.6-qzrsmlmxgad7ckz7ob2xnz3ir7wchclp/lib -L/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/ncurses-6.0-vu7i5i6w5d22kkb3dn2i3xm3bnwmayvv/lib -L/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/zlib-1.2.8-qsdhmcv7nngu2djw7jvldxzb667sxxai/lib -L/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/openssl-1.0.2j-nf26w5xnlarn7b4tezfsqpndhsx6coa5/lib -L/usr/local/lib -L. -lpython2.7 -o build/lib.linux-x86_64-2.7/cPickle.so
...
copying build/lib.linux-x86_64-2.7/cPickle.so -> /calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/python-2.7.12-yotbp4vq4sikgho7gxdoadaf7j3kr2zd/lib64/python2.7/lib-dynload
```

The only thing which catches my eye is that `lib64` in logs during copying, whereas in Spack `package.py` of Python has:
```
    def python_lib_dir(self):
        return join_path('lib', 'python{0}'.format(self.version.up_to(2)))
```

[build.out.zip](https://github.com/LLNL/spack/files/582818/build.out.zip)

-----

One more things from the build environment of PETSc is the following
```
PYTHONHOME=/calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/python-2.7.12-yotbp4vq4sikgho7gxdoadaf7j3kr2zd
PYTHONPATH=
PYTHONSTARTUP=/etc/pythonstart
```
I have no idea what that `PYTHONSTARTUP` means, but it looks like it comes from the environment set up by `OpenSuse` (shouldn't environment be sanitized completely?! I thought so...). Maybe that is the source of the problems.

**Comment by alalazo** [comment]
> I have no idea what that PYTHONSTARTUP means, but it looks like it comes from the environment set up by OpenSuse (shouldn't environment be sanitized completely?! I thought so...). Maybe that is the source of the problems.

@davydden Try this then (if you didn't already) to see if you can narrow down the problem:
```console
$ unset PYTHONSTARTUP
$ /calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/python-2.7.12-yotbp4vq4sikgho7gxdoadaf -c 'import cPickle'
```
`PYTHONSTARTUP` is a variable that may contain a path to a python file that will be executed when you start an interpreter. More info [here](https://docs.python.org/2.7/using/cmdline.html#envvar-PYTHONSTARTUP).

**Comment by alalazo** [comment]
@davydden The other thing that at a first glance seems wrong is `lib64`, as you correctly spotted. From python docs:

> **PYTHONHOME:** Change the location of the standard Python libraries. By default, the libraries are searched in `prefix/lib/pythonversion` and `exec_prefix/lib/pythonversion`, where prefix and exec_prefix are installation-dependent directories, both defaulting to /usr/local.
>
>When PYTHONHOME is set to a single directory, its value replaces both prefix and exec_prefix. To specify different values for these, set PYTHONHOME to prefix:exec_prefix.

So it seems `lib` is expected from a python installation, not `lib64`.

**Comment by alalazo** [comment]
@davydden Following on the previous message, **if** you directory structure looks like:
```console
$ ls /calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/python-2.7.12-yotbp4vq4sikgho7gxdoadaf
bin  include  lib64  share
```
you may try to see if:
```console
$ ln -s <python-prefix>/lib64  <python-prefix>/lib
$ /calculate/spack/opt/spack/linux-opensuse13-x86_64/gcc-4.8/python-2.7.12-yotbp4vq4sikgho7gxdoadaf/bin/python -c 'import cPickle'
```
changes the behavior seen by @christianbaensch .

**Comment by davydden** [comment]
unsetting `PYTHONSTARTUP` did not help, unfortunately.

---

Spack's python on OpenSuse has the following directory structure:
```
├── bin
├── include
│   └── python2.7
├── lib
│   └── python2.7
│       ├── bsddb
│       │   └── test
│       ├── compiler
│       ├── ctypes
│       │   ├── macholib
│       │   └── test
│       ├── curses
│       ├── distutils
│       │   ├── command
│       │   └── tests
│       ├── email
│       │   ├── mime
│       │   └── test
│       │       └── data
│       ├── encodings
│       ├── ensurepip
│       │   └── _bundled
│       ├── hotshot
│       ├── idlelib
│       │   ├── Icons
│       │   └── idle_test
│       ├── importlib
│       ├── json
│       │   └── tests
│       ├── lib-tk
│       │   └── test
│       │       ├── test_tkinter
│       │       └── test_ttk
│       ├── lib2to3
│       │   ├── fixes
│       │   ├── pgen2
│       │   └── tests
│       │       └── data
│       │           └── fixers
│       │               └── myfixes
│       ├── logging
│       ├── multiprocessing
│       │   └── dummy
│       ├── plat-linux2
│       ├── pydoc_data
│       ├── site-packages
│       ├── sqlite3
│       │   └── test
│       ├── test
│       │   ├── audiodata
│       │   ├── capath
│       │   ├── cjkencodings
│       │   ├── data
│       │   ├── decimaltestdata
│       │   ├── imghdrdata
│       │   ├── subprocessdata
│       │   ├── tracedmodules
│       │   └── xmltestdata
│       ├── unittest
│       │   └── test
│       ├── wsgiref
│       └── xml
│           ├── dom
│           ├── etree
│           ├── parsers
│           └── sax
├── lib64
│   ├── pkgconfig
│   └── python2.7
│       ├── config
│       └── lib-dynload
└── share
    └── man
        └── man1
```
so it has both `lib` and `lib64`. `cPickle.so` is located in `/lib64/python2.7/lib-dynload/cPickle.so`. Whereas on my `Ubuntu` it is in `/lib/python2.7/lib-dynload/cPickle.so`.

So it must be related to some config settings for python...

I will try symlinking that `lib-dynload` folder to be within `lib` instead of `lib64`.

**Comment by davydden** [comment]
> I will try symlinking that lib-dynload folder to be within lib instead of lib64.

this worked! I will create a PR to do that extra step during installation...
